Add jsonpath support for config set

mpuckett159 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR intends to add jsonpath as a support method of querying settings to be changed by the `config set` command.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1093

#### Special notes for your reviewer:
This is a continuation of https://github.com/kubernetes/kubernetes/pull/107305, which will be closed.

#### Does this PR introduce a user-facing change?
```release-note
kubectl config (un)set now supports the usage of jsonpath queries to find and set values.
```

ardaguclu commented:
/ok-to-test

mpuckett159 commented:
Small update, still not all the way there, added a property to array and filter nodes similar to field nodes that stores the whole array/filter text in it to make traversing the jsonpath query much easier. At this point cluster and context has full coverage minus setting any types that aren't strings, I haven't implemented that just yet.

Auth-provider and exec can not be set currently either because the jsonpath query does not return an empty value for currently unknown reasons if the auth-provider/exec key doesn't already exist.

If it does exist in the config already we have coverage on exec for the most part, name of the auth-provider, but none for the auth-provider config because I haven't implemented something to deal with setting a map[string]string type yet, though that should be easy enough.

Something else to do is implement a way for a filter node that returns nothing to create the filtered value it's looking for if it's a == filter so we can set things like `{.clusters[?(@.name=="cluster-name")].cluster.server` will create the `cluster-name` cluster if it doesn't already exist.

mpuckett159 commented:
Setting new named structs by filter nodes if they don't already exist works now.

mpuckett159 commented:
Some of the auth-provider and exec keys are settable now, specifically the auth-provider name, and any of the string keys for exec.

Code is a complete mess but I'll clean it up and make it presentable when I'm done.

fedebongio commented:
/assign @seans3 
/triage accepted

mpuckett159 commented:
Quick example of some of the powerful stuff this will unlock:
`kubectl config set '{.users[*].user.username}' "testwildcard"`

The above will set the username key for all the users in the config to `testwildcard`. I can't figure out a real use case for having multiple top level nodes (e.g. `{.key1}{.key2}` that would be set to the same values) but I'm sure someone will figure out a way to use that.

mpuckett159 commented:
/test pull-kubernetes-unit

mpuckett159 committed:
Update jsonpath nodes and parser

I need an easy way to access the actual string value of the filter and array node types, so here I add a new property to each and update the parser to set these new values appropriately.

mpuckett159 committed:
Update unit tests to take into account new properties

mpuckett159 committed:
Add case handling for if a user passes a jsonpath string to set

mpuckett159 committed:
Add case handling for if a user passes a jsonpath string to unset

mpuckett159 committed:
Add unit tests for set and unset with jsonpath

mpuckett159 committed:
Add integration test for new ability to set kubeconfig by jsonpath

mpuckett159 commented:
/retest

mpuckett159 committed:
ci fixes

mpuckett159 commented:
/retest

mpuckett159 committed:
Add deprecation warning for old dot delimited config set syntax

Plumb genericclioptions streams into set and unset to enable this.

mpuckett159 committed:
Update tests to make use of new streams

mpuckett159 commented:
/retest-required

alexzielenski commented on test/cmd/kubeconfig.sh:
```diff
@@ -43,6 +43,29 @@ run_kubectl_config_set_tests() {
+  kubectl config set '{.clusters[?(@.name == "test-cluster")].cluster.certificate-authority-data}' "$encoded"
+  e_written=$(kubectl config view --raw -o jsonpath='{.clusters[?(@.name == "test-cluster")].cluster.certificate-authority-data}')
+
+  test "$e_written" == "$r_written"
```
I do not see how this tests that functionality of using jsonpath. The expected output is also a result from using the new jsonpath feature. What if both responses were wrong but equal?

mpuckett159 replied:
Thank you for pointing this out, this was a complete misunderstanding of what the original test I copied this from was doing, I'll update that so that it actually tests something useful.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/unset.go:
```diff
@@ -104,6 +135,14 @@ func (o *unsetOptions) complete(cmd *cobra.Command, args []string) error {
+
+	// try to determine if we have a jsonpath from first character of first argument
+	// this should only ever be a { if this is a jsonpath string, if it is not we will try to use the old dot delimited
+	// syntax instead.
```
I am curious why the old implementation is left rather than being removed since jsonpath appears to be a superset of the old style. 

It seems to me you could turn any of the old queries into a json path query simply by prefixing it with "$.". Is that true?

mpuckett159 replied:
Unfortunately it is not possible to simply translate the old syntax to the new syntax. It is definitely possible, however it seems like too much extra effort for something we intend to deprecate eventually anyway.

As an example, a translation of setting a cluster resource's server field would be:

Original - `clusters.test-cluster.server`
jsonpath - `{.clusters[?(@.name=="test-cluster")].cluster.server`

The only way to translate these would be to hard code the fact that the middle step in the path (`test-cluster`) would need to be dropped into the hard coded `[?(@.name=="$")].cluster`, which, while doable, implements some pretty ugly code that we will need to remove later anyway. If we wanted to do this without hardcoding values (which is the only way I'd be willing to do it) we would need to create another set of logic to basically search for the fields that the associate clientcmd types use, the associated clientcmdapiv1 types use, then do a lookup to figure out what field we're doing the "filter" on, which is where the name in `@.name` comes from, and then find the field we need to put in the next node in the jsonpath, which can't really be "gotten" in a controller way and would have to be hard coded to cluster anyway. At best we would split and look backwards at clusters and parse out the singular form of the word which is still pretty icky imo.

tldr jsonpath isn't really a super set but more of a parallel of the old style that does overlap in several areas but unfortunately doesn't allow for a bi-directional mapping between the two.

alexzielenski commented on staging/src/k8s.io/client-go/util/jsonpath/parser.go:
```diff
@@ -392,7 +394,7 @@ Loop:
 			return err
 		}
-		cur.append(newFilter(leftParser.Root, rightParser.Root, value[2]))
+		cur.append(newFilter(leftParser.Root, rightParser.Root, value[2], valueText))
```
I am not clear about why these changes to preserve valueText are necessary? I see it being used for something in the test diff.

Edit: I see now they are used to decompose the parsed texted to then recompose it to again valid jsonpath to parse again. I do not think that is a good idea.

mpuckett159 replied:
The reason I added these new value properties is because it seemed like a better idea to create some convenience properties upstream rather than implement logic downstream in kubectl (or wherever) to reconstruct the value from the other Node properties. This probably seems weird because normally you would expect the jsonpath query to be valid and return what you're looking for, but unfortunately we are not guaranteed that a property in a config file will be set already, and in fact it is important that we're able to work on a completely empty config file meaning nothing is set, for us to be able to replicate the functionality of the old syntax. The only way I could think of to accomplish this was to parse everything out into a node list and just rebuild until we reach the end of the list basically, setting things as we go.

alexzielenski replied:
Would it be possible to add an option to jsonpath to create missing intermediate objects as it walks?

mpuckett159 replied:
I don't think that would be feasible. As you can see in this PR the logic required to accomplish that is complicated and highly specialized to the specific data structure that jsonpath is being used to query.

soltysh replied:
> Would it be possible to add an option to jsonpath to create missing intermediate objects as it walks? Kinda like how allowMissingKeys was added as an option

I'm not sure that's desirable here, given that the primary use-case for JSONPath here is printing only. The proposed approach in this PR seems the most minimal, although I'm still thinking if this is correct approach. Only now I see the limitation of our jsonpath library :disappointed:

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/unset.go:
```diff
@@ -80,21 +88,44 @@ func (o unsetOptions) run(out io.Writer) error {
+		}
+		config = finalConfig
+	} else {
+		if _, err := fmt.Fprintln(o.streams.ErrOut, "Warning: usage of dot delimited path for setting config values is deprecated, please use jsonpath syntax instead."); err != nil {
```
Does it make sense to deprecate the simple dot syntax people are used to? I would expect both to work without issue.

alexzielenski commented on staging/src/k8s.io/client-go/util/jsonpath/node.go:
```diff
@@ -139,12 +139,14 @@ type ParamsEntry struct {
 type ArrayNode struct {
 	NodeType
 	Params [3]ParamsEntry // start, end, step
+	Value  string
```
I do not see this additional field being read anywhere in the diff (or any of the other similarly new `Value` fields)

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/unset.go:
```diff
@@ -33,25 +35,31 @@ import (
 	Unset an individual value in a kubeconfig file.
 
-	PROPERTY_NAME is a dot delimited name where each token represents either an attribute name or a map key.  Map keys may not contain dots.`))
+	PROPERTY_NAME is either a jsonpath query for where the property name will be, or a dot delimited name.`))
```
```suggestion
	PROPERTY_NAME is either a jsonpath query for where the property name will be, or a dot delimited name.`))
```
```suggestion
	PROPERTY_NAME is either a dot-eliminated field path, or a jsonpath query which identifies the properties to act upon`))
```

mpuckett159 replied:
This may be a bit pedantic, but I think that "identifies the properties to act upon" is inaccurate because it is possible that the property the user is trying to set does not exist yet, and so when I say "where the properties will be" I specifically mean that the path is where it will be if it does not already exist, or where it will be found if it does exist. Agree that I should not say "property name" and just "properties" is better.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -38,13 +42,17 @@ type setOptions struct {
 	propertyValue string
 	setRawBytes   cliflag.Tristate
+	jsonPath      bool
+	deduplicate   bool
```
This `deduplicate` feature seems like it might be out of scope for this PR?

mpuckett159 replied:
Yeah this may not be necessary really. I originally wrote the functions that support this flag to allow me to extend the existing syntax, which didn't support the idea of being able to set a specific value in an array or filtering results, but this implementation using jsonpath now supports that, so this may be pointless I guess. If others could chime in if they think this is necessary or not I'd appreciate it.

It's also a bit of a bad name for the flag because the it uses the `k8s.io/apimachinery/pkg/util/sets` which also sorts in addition to deduplicating.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/unset.go:
```diff
@@ -33,25 +35,31 @@ import (
+		kubectl config unset contexts.foo.namespace
+
+		# Unset cluster server using jsonpath
+		kubectl config unset '{.clusters[?(@.name=="cluster-0")].cluster.server}' --jsonpath`)
```
Is this the correct usage? In `set.go` it appears the `--jsonpath` flag is not needed?

mpuckett159 replied:
Whoops thanks for the catch.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -141,6 +202,311 @@ func (o setOptions) validate() error {
 	return nil
 }
 
+// modifyConfigJson: modifyConfig, but for JSON path
```
Could you write more about what this function does? `modifyConfig` does not have documentation either so this comment is not very helpful as documentation.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
I find it very odd that the above switch case composes strings which are then parsed again as json path. Isn't the user input expected to be valid jsonpath?

mpuckett159 replied:
The users input is expected to be valid jsonpath, however, the existing config is not expected to have a value at the location of the user provided jsonpath, and so we must basically "walk" the provided jsonpath query and create the necessary objects as we go before finally setting the value the user is specifically asking to set. That way if a user were to have an empty config file, they can use set on its own to build a config file from scratch. The same is desired to be true for things like creating a new cluster or context without having to do individual commands for setting every step in the path. Otherwise we would need to set every step in the path with individual commands, e.g.:

`set {.clusters[?(@.name=="test-cluster")]`
`set {.clusters[?(@.name=="test-cluster")].cluster.server`

Instead of just simply passing the last command.

soltysh replied:
I'm confused just like Alex, so you're saying that when you normally walk using for example `FindResults` it won't create those nodes, but when you manually work it will? Assuming that in both cases you have `AllowMissingKeys` set, it's rather surprising.

mpuckett159 replied:
Yeah so basically if the key doesn't exist it will just return an empty slice of an empty slice of results, and it will not create them. The `AllowMissingKeys` setting simply specifies that no error is returned when you search for something and it isn't there. The in tree library for jsonpath stuff is only for looking up values and not for writing them in any way so this is a bit of a cludge to make that possible without having to substantially change the way our jsonpath library actually works.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -141,6 +202,311 @@ func (o setOptions) validate() error {
+	// Create jsonpath parser to use throughout the function
+	jsonPath := jsonpath.New("Value Getter").AllowMissingKeys(true)
+
+	jsonPathParser, err := jsonpath.Parse("Node Parser", propertyName)
```
Why not call `FindResults` on this instance of `jsonPathParser` instead of the complicated logic below?

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
What is this loop doing? I am having trouble following what "work on the results" means in this context, and why the `innerJsonPath` objects are necessary?

mpuckett159 replied:
This was something I waffled on a bit because it seems burdensome and overly complicated to support a feature of jsonpath that almost no one will probably use in this situation.

Results are returned as `[][]reflect.Value`. This is to support basically having multiple jsonpath's within the jsonpath string, e.g. `{.key1}{.key2}` is valid jsonpath, and the outer slice contains the results for the two separate path's contained within the `{}`s. Then, the inner results are the values that those paths actually return. I honestly can't envision a situation where setting several disparate keys to the same value would be required or helpful, but I didn't want to assume so I decided to just support it.

Fair point though I need to make the comments clearer. Most of these were just "for me" comments that I intended to update later and forgot about.

alexzielenski reviewed (commented):
Thanks for your contribution! Apologies if this is not ready for review, but I thought I would give it a pass.

This is a great change which closes a functionality gap in kubectl `set/unset`. However, I found some confusing aspects of your implementation I am hoping you can clear up.

mpuckett159 committed:
Fix typo in help text of unset

mpuckett159 committed:
Update comments

mpuckett159 committed:
Implement more safety checking

So users don't have to look at panics when they don't enter a perfect jsonpath.

mpuckett159 committed:
make e2e test actually test something useful

mpuckett159 committed:
Clarify language on help text for set/unset

mpuckett159 committed:
Make sure unset doesn't unnecessarily create stuff

mpuckett159 committed:
Fix spelling error

mpuckett159 commented:
/remove-lifecycle stale

soltysh reviewed (commented):
I left some comments, I'm still thinking about that write part of our jsonpath library, I kinda feel like I've put you on a mine with that idea :disappointed: - sorry for that.

mpuckett159 committed:
Move modifyConfigJson and supporting functions to new file

	modified:   staging/src/k8s.io/kubectl/pkg/cmd/config/set.go
	new file:   staging/src/k8s.io/kubectl/pkg/cmd/config/set_jsonpath.go

mpuckett159 committed:
Address feedback

Update options to o
Move validate action to public cobra function
Create modifyConfig functions for readability
Simplify jsonPath bool set

mpuckett159 committed:
go fmt update

I really don't like this formatting but w/e I guess

mpuckett159 commented:
/test pull-kubernetes-integration

dims commented:
If you still need this PR then please rebase, if not, please close the PR
